src/interchange/src/envelopes.rs

```diff
@@ -43,72 +48,66 @@ where
-                        // Slurp it into a vector, so we can sort that by timestamps.
+                        let mut cursor = batch.cursor();
                         while cursor.key_valid(&batch) {
                             let k = cursor.key(&batch);
+
+                            // Partition updates into retractions (befores)
+                            // and insertions (afters).
                             while cursor.val_valid(&batch) {
-                                let val = cursor.val(&batch);
+                                let v = cursor.val(&batch);
                                 cursor.map_times(&batch, |&t, &diff| {
-                                    buf.push((t, k.as_ref(), diff, val));
+                                    let update = (t, v, usize::cast_from(diff.unsigned_abs()));
+                                    if diff < 0 {
+                                        befores.push(update);
+                                    } else {
+                                        afters.push(update);
+                                    }
                                 });
                                 cursor.step_val(&batch);
                             }
-                            cursor.step_key(&batch);
-                        }
-                        // For each (timestamp, key) we've seen,
-                        // we need to generate a series of (before, after) pairs.
-                        //
-                        // Steps to do this:
-                        // (1) sort by ts, diff (so that for each ts, diffs appear in ascending order).
-                        // (2) in each (ts, key), find the point where negative and positive diffs meet, using binary search.
-                        // (3) The (ts, entry, diff) elements before that point will go into "before" fields in DiffPairs;
-                        //     the ones after that point will go in "after".
-
-                        // Step (1) above)
-
-                        // Because `sort_by_key` is stable, it will not reorder equal elements. Therefore, elements with the same
-                        // key will stay grouped together.
-                        buf.sort_by_key(|(t, _k, diff, _row)| (*t, *diff));
-                        for ((t, k), group) in &buf
-                            .into_iter()
-                            .group_by(|(t, k, _diff, _row)| (*t, k.clone()))
-                        {
-                            let mut out = vec![];
-                            let elts: Vec<(G::Timestamp, Option<&Row>, Diff, &Row)> =
-                                group.collect();
-                            // Step (2) above
-                            let pos_idx = elts
-                                .binary_search_by(|(_t, _k, diff, _row)| diff.cmp(&0))
-                                .expect_err("there should be no zero-multiplicity entries");
```

Comment:
I curiously, I was unable to trigger this even in the old code -- timely will panic with some other panic text if there is a diff = 0 or if a record is both inserted and deleted at the same timestamp.

Reply:
Yeah, that makes sense; the `expect` here is asserting an invariant that is upheld by earlier layers in the system.